yapapi/script/__init__.py

```diff
@@ -0,0 +1,172 @@
+        batch: List[BatchCommand] = []
+        for cmd, _future in self._commands:
+            batch.append(cmd.evaluate(self._ctx))
+        return batch
+
+    async def _after(self):
+        """Hook which is executed after the script has been run on the provider."""
+        for cmd, _future in self._commands:
+            await cmd.after(self._ctx)
+
+    async def _before(self):
+        """Hook which is executed before the script is evaluated and sent to the provider."""
+        if not self._ctx._started and self._ctx._implicit_init:
+            loop = asyncio.get_event_loop()
+            self._commands.insert(0, (Deploy(), loop.create_future()))
+            self._commands.insert(1, (Start(), loop.create_future()))
+            self._ctx._started = True
+
+        for cmd, _future in self._commands:
+            await cmd.before(self._ctx)
+
+    def _set_cmd_result(self, result: CommandExecuted) -> None:
+        cmd, future = self._commands[result.cmd_idx]
+        future.set_result(result)
+
+    def add(self, cmd: Command) -> Awaitable[CommandExecuted]:
+        loop = asyncio.get_event_loop()
+        future_result = loop.create_future()
+        self._commands.append((cmd, future_result))
+        return future_result
+
+    def deploy(self) -> Awaitable[CommandExecuted]:
+        """Schedule a Deploy command on the provider."""
+        return self.add(Deploy())
+
+    def start(self, *args: str) -> Awaitable[CommandExecuted]:
+        """Schedule a Start command on the provider."""
+        return self.add(Start(*args))
+
+    def terminate(self) -> Awaitable[CommandExecuted]:
+        """Schedule a Terminate command on the provider."""
+        return self.add(Terminate())
+
+    def send_json(self, data: dict, dst_path: str) -> Awaitable[CommandExecuted]:
+        """Schedule sending JSON data to the provider.
+
+        :param data: dictionary representing JSON data to send
+        :param dst_path: remote (provider) destination path
+        """
+        return self.add(SendJson(data, dst_path))
+
+    def send_bytes(self, data: bytes, dst_path: str) -> Awaitable[CommandExecuted]:
+        """Schedule sending bytes data to the provider.
+
+        :param data: bytes to send
+        :param dst_path: remote (provider) destination path
+        """
+        return self.add(SendBytes(data, dst_path))
+
+    def send_file(self, src_path: str, dst_path: str) -> Awaitable[CommandExecuted]:
+        """Schedule sending a file to the provider.
+
+        :param src_path: local (requestor) source path
+        :param dst_path: remote (provider) destination path
+        """
+        return self.add(SendFile(src_path, dst_path))
+
+    def run(
+        self,
+        cmd: str,
+        *args: str,
+        env: Optional[Dict[str, str]] = None,
+        stderr: CaptureContext = CaptureContext.build(mode="stream"),
+        stdout: CaptureContext = CaptureContext.build(mode="stream"),
```

Comment:
Those defaults are copied in `yapapi.script.command.Run`. This is a good potential for future bugs because this might always be kept synchronized.

I would rather have those default to `None` and
```
kwargs = []
if stderr is not None:
    kwargs[stderr] = stderr
...
return self.add(Run(cmd, *args, **kwargs)).

Reply:
(also I have some general note)

Reply:
Agreed, let's keep the defaults in one place.

Reply:
Updated in 37f3dfe